Adds firebots to the robotics arsenal

Lett1 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

Adds a new type of bot: the Firebot!

![https://i.imgur.com/rd2Jk5R.png](https://i.imgur.com/rd2Jk5R.png)

This lil guy will scoot around and tries his best to extinguish burning people and hotspots around him.
If emagged he will knock people over and wets the floor under em for extra salt.

Sprites by Nemvar on discord. Sfx by generic windows tts.

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
Firebots are supposed to be used by mapmakers/people in highly flammable enviroments, like chemistry or toxins. The current way of fighting fires is to either drain the air out of the room or to go after it with extinguishers. The bot is supposed to support the later, he is not supposed to be a cure all for dangerous plasma fires.

## Changelog
:cl:
add: Added a new type of bot: the firebot.
add: Firebots can be constructed out of extinguisher + borg arm > red fire helmet > prox sensor
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Lett1 committed:
initial firebot commit

Lett1 committed:
add sprites for the firebot

Lett1 committed:
firebot.dm: More work on the firebot, implemented emag behaviour and better extinguishing
construction.dm: Added firebot assembly

Lett1 committed:
extinguisher.dm: Added support for the firebot assembly

Lett1 committed:
fix build error

davethwave commented:
Seems like a ok idea.

habader commented:
Add a slime murdering option

Lett1 committed:
commit missing dme

tralezab commented:
the emagged version just knocks people over? nothing to do with fire?

Lett1 commented:
> the emagged version just knocks people over? nothing to do with fire?

I think setting people on fire or making fire spread more would be to dangerous for something like that since fire can kill rather quickly.

ShizCalev commented:
doesn't drop the fire helmet? :(

anconfuzedrock commented:
why firesuit helmets? aren't those like rare as fuck? unless you mean just red helmets from emergency lockers in which case I swear to god if I start opening fire closets and all the helmets are gone I'm going to gut somebody

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,295 @@
+
+	new /obj/item/assembly/prox_sensor(Tsec)
+
+	var/turf/T = get_turf(loc)
```
get_turf(src)

Lett1 replied:
Fixed, also drops the helmet now.

Lett1 committed:
now drops the helmet again, also added a message for knocking people over

Lett1 commented:
> why firesuit helmets? aren't those like rare as fuck? unless you mean just red helmets from emergency lockers in which case I swear to god if I start opening fire closets and all the helmets are gone I'm going to gut somebody

Yeah I mean those, gonna edit the text.

Lett1 committed:
fix previous commit, thats what you get for pushing without building first

Lett1 committed:
firebot now deploys firefighting foam when under heat instead of spraying wildly

Lett1 committed:
fix typos, move defines to top of file and adds undefines to the end

MrPerson commented:
>I swear to god if I start opening fire closets and all the helmets are gone I'm going to gut somebody

Peak SS13 gameplay right there. Maybe we should spawn some of the lockers missing stuff?

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,307 @@
+/mob/living/simple_animal/bot/firebot/explode()
+	on = FALSE
+	visible_message("<span class='boldannounce'>[src] blows apart!</span>")
+	var/atom/loc = drop_location()
```
don't name this var loc. :(

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,307 @@
+	visible_message("<span class='boldannounce'>[src] blows apart!</span>")
+	var/atom/loc = drop_location()
+
+	new /obj/item/assembly/prox_sensor(loc)
```
.

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,307 @@
+	var/atom/loc = drop_location()
+
+	new /obj/item/assembly/prox_sensor(loc)
+	new /obj/item/clothing/head/hardhat/red(loc)
```
.

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,307 @@
+		theturf.MakeSlippery(TURF_WET_WATER, min_wet_time = 10 SECONDS, wet_time_to_add = 5 SECONDS)
+
+	if(prob(50))
+		drop_part(robot_arm, loc)
```
no.

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,307 @@
+	new /obj/item/assembly/prox_sensor(loc)
+	new /obj/item/clothing/head/hardhat/red(loc)
+
+	var/turf/T = get_turf(loc)
```
get_turf(src) still

OverlyGeneric commented:
Goonstation's **Leaner** Cousin

ATH1909 commented:
Idea here: Instead of having the firebot knock you down directly and then dump water on the ground, it should dump water on the ground and then swap tiles with you (even though bots don't normally swap tiles with mobs, IIRC), slipping you like how a humanoid with a fire extinguisher would.

OverlyGeneric commented:
> Idea here: Instead of having the firebot knock you down directly and then dump water on the ground, it should dump water on the ground and then swap tiles with you (even though bots don't/can't normally swap tiles with mobs, IIRC), slipping you like a humanoid with a fire extinguisher would.

I still think the thing should be able to wail onto the poor bastard that steps into the red bull's way. 

However, you could make this possible by having it move onto the same tile as the player like most other friendly mobs do when they randomly stumble around.

GunHog commented:
Slipping people is the cleanbot's thing, though...

PKPenguin321 commented:
\>emagging it doesn't make it shoot fire
Lame!

OverlyGeneric commented:
> >emagging it doesn't make it shoot fire
> >emagging doesn't make it shoot snow and freeze people to death as it beats everyone with its robust body.
> >lamer

dr-bee commented:
Having a firebot that doesnt shoot fire somehow is a waste of a good pun. Also red fire helmets arent mass printable making a hard limit to the amount of firebots that can be built. Consider another item or make the helmets printable at an autolathe or order-able from cargo.

Lett1 committed:
renamed loc to drop_location for clarity

Lett1 commented:
> Having a firebot that doesnt shoot fire somehow is a waste of a good pun. Also red fire helmets arent mass printable making a hard limit to the amount of firebots that can be built. Consider another item or make the helmets printable at an autolathe or order-able from cargo.

You can order a firefighting crate from cargo, it contains two helmets.

Lett1 committed:
fix drop location for real this time

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+	visible_message("<span class='boldannounce'>[src] blows apart!</span>")
+
+	new /obj/item/assembly/prox_sensor(src.loc)
+	new /obj/item/clothing/head/hardhat/red(src.loc)
```
use drop_location(). You had this all right the first time with `Tsec`, not sure why you keep changing it. Refer to the below code.

https://github.com/tgstation/tgstation/blob/a9ca93b60dbf7b9f0fc5cbe94c406a2126312fee/code/modules/mob/living/simple_animal/bot/floorbot.dm#L361-L379

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+	on = FALSE
+	visible_message("<span class='boldannounce'>[src] blows apart!</span>")
+
+	new /obj/item/assembly/prox_sensor(src.loc)
```
.

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+		theturf.MakeSlippery(TURF_WET_WATER, min_wet_time = 10 SECONDS, wet_time_to_add = 5 SECONDS)
+
+	if(prob(50))
+		drop_part(robot_arm, src.loc)
```
.

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+		return
+	if(IsStun() || IsParalyzed())
+		icon_state = "firebots1"
+		return
```
this return isn't doing anything

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+	if(is_burning(scan_target))
+		if((detected_cooldown + DETECTED_VOICE_INTERVAL) < world.time)
+			speak("Fire detected!")
+			playsound(src.loc, "sound/voice/firebot/detected.ogg", 50, 0)
```
```suggestion
			playsound(src, "sound/voice/firebot/detected.ogg", 50)
```

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+			src.do_attack_animation(M)
+			visible_message("<span class='danger'>[src] rams into [M] knocking them over!</span>")
+
+			var/turf/T = get_turf(M.loc)
```
```suggestion
			var/turf/T = get_turf(M)
```

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+			var/mob/living/M = target_fire
+			M.Knockdown(3)
+			src.do_attack_animation(M)
+			visible_message("<span class='danger'>[src] rams into [M] knocking them over!</span>")
```
```suggestion
			visible_message("<span class='danger'>[src] rams into [M] knocking [M.p_them()] over!</span>")
```

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+				playsound(src.loc, "sound/voice/firebot/stopdropnroll.ogg", 50, 0)
+			else
+				speak("Extinguishing!")
+				playsound(src.loc, "sound/voice/firebot/extinguishing.ogg", 50, 0)
```
```suggestion
				playsound(src, "sound/voice/firebot/extinguishing.ogg", 50)
```

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+		if((speech_cooldown + SPEECH_INTERVAL) < world.time)
+			if(ishuman(target_fire))
+				speak("Stop, drop and roll!")
+				playsound(src.loc, "sound/voice/firebot/stopdropnroll.ogg", 50, 0)
```
```suggestion
				playsound(src, "sound/voice/firebot/stopdropnroll.ogg", 50, 0)
```

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+		"Keep it cool." = 'sound/voice/firebot/keepitcool.ogg')
+		var/message = pick(messagevoice)
+		speak(message)
+		playsound(loc, messagevoice[message], 50, 0)
```
```suggestion
		playsound(src, messagevoice[message], 50)
```

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+		if(user)
+			to_chat(user, "<span class='danger'>[src] buzzes and beeps.</span>")
+		audible_message("<span class='danger'>[src] buzzes oddly!</span>")
+		playsound(src, "sparks", 75, 1)
```
```suggestion
		playsound(src, "sparks", 75, TRUE)
```

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,306 @@
+
+/mob/living/simple_animal/bot/firebot/Topic(href, href_list)
+	if(..())
+		return 1
```
```suggestion
		return TRUE
```

ShizCalev commented on code/game/objects/items/extinguisher.dm:
```diff
@@ -229,3 +232,12 @@
+		qdel(src)
+		user.put_in_hands(new /obj/item/bot_assembly/firebot)
+	else
+		..()
```
new line at the end.

Lett1 committed:
extinguisher.dm: Add newline to eof
firebot.dm: Make explode() use drop_location properly (for real this time), text fixes, removed improper use of .loc

Lett1 commented:
Sorry you have to put up with my shit @ShizCalev, lets hope it works this time.

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,308 @@
+		"Keep it cool." = 'sound/voice/firebot/keepitcool.ogg')
+		var/message = pick(messagevoice)
+		speak(message)
+		playsound(loc, messagevoice[message], 50)
```
```suggestion
 		playsound(loc, messagevoice[message], 50)
```

ShizCalev committed:
Update code/modules/mob/living/simple_animal/bot/firebot.dm

Co-Authored-By: Lett1 <kilian2311@gmail.com>

ATH1909 commented:
> Slipping people is the cleanbot's thing, though...

Oh, true.

Maybe have it shoot out frozen water vapor instead of the room temperature(?) water it normally shoots out, freezing large areas of the station?

ShizCalev commented on code/modules/mob/living/simple_animal/bot/firebot.dm:
```diff
@@ -0,0 +1,308 @@
+		if(emagged == 2) // When emagged, knock people down and wet the floors under em
+			var/mob/living/M = target_fire
+			M.Knockdown(3)
+			src.do_attack_animation(M)
```
remove src

anconfuzedrock commented:
(((I WAS SERIOUS ABOUT THE FIREFIGHTER HELMETS A CLEANBOT ONLY NEEDS A BUCKET A MEDIBOT ONLY NEEDS A MEDKIT FOR THE LOVE OF GOD AND ALL THAT IS HOLY DO NOT MAKE IT TAKE FIRE HELMETS THE SPRITE DOESN'T EVEN LOOK GOOD)))

Lett1 commented:
Changed how the emagged one works: it now sprays fire (cfl3) onto people.

Lett1 committed:
update emag behaviour to use cfl3

Lett1 committed:
Merge branch 'firebot' of https://github.com/Lett1/tgstation into firebot

TerraGS commented on code/game/objects/items/extinguisher.dm:
```diff
@@ -229,3 +232,12 @@
+		qdel(src)
+		user.put_in_hands(new /obj/item/bot_assembly/firebot)
+	else
+		..()
```
This part should be in construction.dm to keep it consistent with medbots and floorbots.

Lett1 replied:
Didn't know that you could craft em like this, added now.

Lett1 commented:
> Also needs a recipe in /code/modules/crafting/recipes.dm like the other bots.

Didn't know you could craft em like this, added it.

Lett1 committed:
recipes.dm: Add crafting recipe for firebot
extinguisher.dm + construction.dm: Move extinguisher on_attack code to construction.dm

DapperMola commented:
should have emagged lines like the cleanbot
"Burn, baby burn."
"Is it getting hot in here? Or is it just you?"
"Smoking kills."
